scroll-search.lic

```diff
@@ -0,0 +1,533 @@
+	# @owned_scrolls = { }
+	# UserVars.scroll_counter = 0
+
+	$settings = get_settings()
+	$debug_mode = $settings.debug_mode
+
+	echo("Settings debug mode is: #{$debug_mode}") if $debug_mode
+	if $settings.nil? || $settings.empty?
+	  echo("Settings was nil!") if $debug_mode
+	end
+
+	arg_definitions = 
+	[
+	  [
+		{ name: 'search', regex: /^sea.*/i, description: 'Searches given container and labels any unlabeled scrolls. ' },
+		{ name: 'container', regex: /\w+/i, description: 'The container to search. ' }
+	  ],
+	  [
+		{ name: 'list', regex: /^lis.*/i, optional: false, description: 'Lists all of the scrolls that have been categorized after a SEARCH. ' },
+	  ]
+	]
+
+	args = parse_args(arg_definitions)
+
+	unless args.search || args.list
+	  display_args(arg_definitions)
+	  return
+	end
+
+	pause
+
+	if args.search
+	  if args.container.nil?
+		echo_message("SPECIFY THE CONTAINER TO SEARCH!")
+		return
+	  else
+		search_container(args.container)
+	  end
+	elsif args.list
+	  if @owned_scrolls.length <= 0
+		echo_message("NO SCROLLS HAVE BEEN ADDED! SEARCH YOUR CONTAINERS FIRST!")
+	    return
+	  end
+
+	  echo("Total scroll count is: #{@owned_scrolls.length}") if $debug_mode
+	  display_scrolls()
+	else
+	  echo_message("INVALID ACTION")
+	end
+  end
+
+  def search_container(container)
+	# return if invalid container
+
+	unless exists?(container)
+	  echo_message("INVALID CONTAINER")
+	  return
+	end
+
+	# Empty your hands
+	empty_hands
+
+	# Get long name of container
+	$container_full = full_name(container)
+	if $container_full =~ /Undefined/i
+	  echo("Was unable to retrieve the full name of this container...") if $debug_mode
+	  $container_full = container
+	end
+
+	$container = get_noun($container_full)
+
+	# iterate over all scroll types
+	scroll_types = $settings.scroll_nouns
+
+	if scroll_types.empty? || scroll_types.nil?
+	  echo_message("NO SCROLL TYPTES SET! SETUP YOUR YAML TO INCLUDE VALID \"scroll_types\".")
+	  return
+	end
+
+	echo("Scroll type count is: #{scroll_types.length}.") if $debug_mode
+
+	# Clear current scroll listing from owned scrolls if
+	# user is searching an existing container
+	if @owned_scrolls[$container_full]
+	  echo("Clearing scroll array for this container, since it was previously searched --- #{$container_full}!") if $debug_mode
+	  @owned_scrolls[$container_full] = []
+	end
+	
+	# reset scroll count
+	@scroll_count = 0
+	
+	scroll_types.each do |scroll|
+	  echo "Searching for scrolls of type #{scroll}." if $debug_mode
+
+	  scroll_list = get_scrolls(scroll)
+	  next if scroll_list.empty?
+
+	  look_scrolls(scroll_list)
+	end
+
+	# Save scrolls to user variables.
+	if UserVars.owned_scrolls.nil? && @owned_scrolls.nil?
+	  echo("Creating an empty hash for vars!") if $debug_mode
+	  UserVars.owned_scrolls = {}
+	  return
+	end
+	
+	existing_hash = UserVars.owned_scrolls ? UserVars.owned_scrolls.clone : @owned_scrolls.clone
+	new_hash = @owned_scrolls
+	existing_hash.merge!(new_hash) 
```

Comment:
What does this `merge!` accomplish? As far as I can tell, `existing_hash` is not used (and `new_hash` is not modified).